py_async_grpc_prometheus/prometheus_async_server_interceptor.py

```diff
@@ -118,30 +97,81 @@ def new_behavior(request_or_iterator, servicer_context):
               _LOGGER.error(e)
             if response_or_iterator is None:
               return response_or_iterator
-            return behavior(request_or_iterator, servicer_context)
+            return await behavior(request_or_iterator, servicer_context)
           raise e
 
-      return new_behavior
+      async def new_response_stream_behavior(request_or_iterator, servicer_context):
+        try:
+          start = default_timer()
+          grpc_type = grpc_utils.get_method_type(request_streaming, response_streaming)
+          
+          try:
+            self._metrics["grpc_server_started_counter"].labels(
+                  grpc_type=grpc_type,
+                  grpc_service=grpc_service_name,
+                  grpc_method=grpc_method_name).inc()
+            if request_streaming:
+              request_or_iterator = grpc_utils.wrap_iterator_inc_counter(
+                  request_or_iterator,
+                  self._metrics["grpc_server_stream_msg_received"],
+                  grpc_type,
+                  grpc_service_name,
+                  grpc_method_name)
 
-    optional_any = self._wrap_rpc_behavior(continuation(handler_call_details), metrics_wrapper)
+            async for obj in behavior(request_or_iterator, servicer_context):
+              self._metrics["grpc_server_stream_msg_sent"].labels(
+                grpc_type=grpc_type,
+                grpc_service=grpc_service_name,
+                grpc_method=grpc_method_name).inc()
+              yield obj
+
+            self.increase_grpc_server_handled_total_counter(grpc_type,
+                                                            grpc_service_name,
+                                                            grpc_method_name,
+                                                            self._compute_status_code(
+                                                                servicer_context).name)
+          
+          except grpc.RpcError as exc:
+            self.increase_grpc_server_handled_total_counter(
+                grpc_type,
+                grpc_service_name,
+                grpc_method_name,
+                self._compute_error_code(exc).name
+            )
+            raise exc
+
+          finally:
+            self._increase_grpc_server_handled_latency(grpc_type, grpc_service_name, grpc_method_name, start)
+        
+        except Exception as e:
+          if self._skip_exceptions:
+            if self._log_exceptions:
+              _LOGGER.error(e)
+            async for obj in behavior(request_or_iterator, servicer_context):
+              yield obj
+          raise e
+      
+      if response_streaming:
```

Comment:
I am a bit confused with the multiple `try/except` blocks.
 Maybe we can wrap some of the parts in functions?

Reply:
Agree - opening an issue for it.